proposals/2071-update-versioning.md

```diff
@@ -0,0 +1,42 @@
+# Proposal: Column lineage endpoint proposal
+
+Author(s): @rnhttr, @wslulciuc, @collado-mike
+
+Created: 20022-09-30
+
+Discussion:
+* [Proposal: Update DatasetVersion versioning #2071](https://github.com/MarquezProject/marquez/issues/2071)
+* [Discrepancy in building DatasetVersionIds and which version is used #1977 ](https://github.com/MarquezProject/marquez/issues/1977)
+* [Dataset currentVersion not in Dataset Versions Listing #1883](https://github.com/MarquezProject/marquez/issues/1883)
+
+## Overview
+
+The current versioning system leads to confusion (e.g. [#1883](https://github.com/MarquezProject/marquez/issues/1883)). DatasetVersion, for example, has a uuid field (of type UUID) and a version field (also of type UUID). In a practical sense, I think these fields are redundant.
+
+Additionally, external data systems might already support dataset versioning (e.g. delta, iceberg). It'd make sense for Marquez to be able support these external versions.
+
+## Proposal
+This proposal aims to simplify the Marquez versioning system and enable new features to support dataset versions supplied by external data systems.
+
+Moving forward, `DatasetVersionRow` and its cousin [`ExtendedDatasetVersionRow`](https://github.com/MarquezProject/marquez/blob/0.26.0/api/src/main/java/marquez/db/models/ExtendedDatasetVersionRow.java#L26) will no longer support a `version` field; rather, this will be dropped and replaced by a nullable `externalVersion`. This `externalVersion` field will be populated if provided, using [OpenLineage's DatasetVersionDatasetFacet facet](https://github.com/OpenLineage/OpenLineage/blob/main/spec/facets/DatasetVersionDatasetFacet.json) to support external dataset versions. `DatasetVersionRow`'s `uuid` field will serve as the version's sole identifier.
```

Comment:
Can you comment on how the _public_ API is impacted (if at all)?

Reply:
Not from a user's perspective as far as I can tell, although there is another `JobVersion` in the public API that will need the `version` field removed. I'll also need to update the various versions in the clients. I'll add all of these to the list of TODOs in the implementation section.

Reply:
Actually, I guess, while the public API will be backward compatible, we'll to expose `externalVersion` to the public API as well.

Reply:
I was thinking specifically about how the URLs for fetching DatasetVersions will work and how we return references to the DatasetVersion. E.g., [this structure](https://github.com/MarquezProject/marquez/blob/main/spec/openapi.yml#L1186-L1202) in the spec references a UUID version, but I don't know if that is going to change or if it will just reference the DatasetVersion's primary key. So what is returned in the [`Run` spec](https://github.com/MarquezProject/marquez/blob/main/spec/openapi.yml#L1655-L1689)'s `inputDatasetVersions` field?
Additionally, the [`getDatasetVersion` spec](https://github.com/MarquezProject/marquez/blob/main/spec/openapi.yml#L201-L218) accepts a `version` parameter. Is that the UUID? or the `externalVersion`?
Personally, I prefer the `externalVersion`, if present, as it introduces compatibility between systems - e.g., a Snowflake dashboard can just _know_ how to construct a link to the Marquez UI for a given dataset@version because the standard is known and consistent.

Reply:
@collado-mike: very good points on backwards compatibility. To ensure we don't introduce any breaking changes, Marquez will continue to reference the `uuid` of `dataset_versions` and set that value accordingly. That is, in the case of `Datasets.currentVersion`, the value will continue to be of type `uuid`, but with the addition of `Datasets.currentExternalVersion` (we can think of this as an alias). More specifically, for `DatasetVersionId`, we will add `Datasets.externalVersion` which can be `null`, but `Datasets.version` will always be set.

I think retrieving a dataset version is less obvious, but we can update the endpoint:

```
GET /namespaces/{namespace}/datasets/{dataset}/versions/{version}
```

where `{version}` can either be a `uuid` or a string type. The parsing on the backend really isn't that complicated and be easy enough to do.

Another option would be we expose a new endpoint, but that seems unnecessary and would prefer being flexible on `{version}` type:

```
GET /namespaces/{namespace}/datasets/{dataset}/versions/{version}/external
```